File: java-example/src/main/java/com/snowflake/example/ElasticStep2Continuous.java
package com.snowflake.example;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestElasticChannel;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.Duration;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

/**
 * Elastic step 2: keep appending while bounding unacknowledged work.
 *
 * <p>The deque stores Future handles, not rows. The SDK owns row buffering,
 * batching, and transient transport retries.
 */
public class ElasticStep2Continuous {
    static final int MAX_PENDING_EVENTS = 10_000;

    static SnowflakeStreamingIngestClient createClient() throws Exception {
        Properties properties = new Properties();
        JsonNode profile = new ObjectMapper().readTree(Files.readAllBytes(
                Paths.get(env("SNOWFLAKE_PROFILE", "profile.json"))));
        profile.fields().forEachRemaining(
                entry -> properties.put(entry.getKey(), entry.getValue().asText()));
        return SnowflakeStreamingIngestClientFactory.tableBuilder(
                "continuous",
                env("SNOWFLAKE_DATABASE", "MY_DATABASE"),
                env("SNOWFLAKE_SCHEMA", "MY_SCHEMA"),
                env("SNOWFLAKE_TABLE", "MY_TABLE"))
                .setProperties(properties)
                .build();
    }

    static String env(String name, String fallback) {
        String value = System.getenv(name);
        return value == null || value.isBlank() ? fallback : value;
    }

    static int waitAndRemoveConfirmedPrefix(Deque<CompletableFuture<Void>> pending)
            throws Exception {
        // An exceptional result is terminal from the SDK's perspective; propagate it.
        pending.peekFirst().get();
        int confirmed = 0;
        // One SDK acknowledgement may complete several consecutive append Futures.
        while (!pending.isEmpty() && pending.peekFirst().isDone()) {
            pending.removeFirst().get();
            confirmed++;
        }
        return confirmed;
    }

    static Map<String, Object> sampleRow(long eventId) {
        return Map.of(
                "EVENT_ID", eventId,
                "C1", eventId,
                "C2", "event-" + eventId);
    }

    public static void main(String[] args) throws Exception {
        long total = Long.parseLong(env("SNOWFLAKE_TEST_ROWS", "10000"));
        SnowflakeStreamingIngestClient client = createClient();
        Deque<CompletableFuture<Void>> pending = new ArrayDeque<>();
        int confirmed = 0;
        boolean completed = false;
        try {
            SnowflakeStreamingIngestElasticChannel channel = client.getElasticChannel();
            for (long eventId = 0; eventId < total; eventId++) {
                // No callback is registered, so the returned Future identifies the append.
                pending.addLast(channel.appendRowWithWait(sampleRow(eventId), null));
                if (pending.size() >= MAX_PENDING_EVENTS) {
                    // Pause intake until at least one acknowledgement slot is released.
                    confirmed += waitAndRemoveConfirmedPrefix(pending);
                }
            }

            // End of input: wait until every accepted append is durably acknowledged.
            while (!pending.isEmpty()) {
                confirmed += waitAndRemoveConfirmedPrefix(pending);
            }
            completed = true;
            System.out.println("Durably acknowledged " + confirmed + " rows");
        } finally {
            client.close(completed, Duration.ofSeconds(60)).get(60, TimeUnit.SECONDS);
        }
    }
}

File: nodejs-example/elastic_step2_continuous.js
#!/usr/bin/env node
/**
 * Elastic step 2: keep appending while bounding unacknowledged work.
 *
 * The array stores Promise handles, not rows. The SDK owns row buffering,
 * batching, and transient transport retries. A rejection that reaches this
 * example is terminal and stops the program.
 */
"use strict";

const streaming = require("snowpipe-streaming");

const MAX_PENDING_EVENTS = 10_000;

async function createClient() {
  return streaming.createTableClient({
    clientName: "continuous",
    dbName: process.env.SNOWFLAKE_DATABASE || "MY_DATABASE",
    schemaName: process.env.SNOWFLAKE_SCHEMA || "MY_SCHEMA",
    tableName: process.env.SNOWFLAKE_TABLE || "MY_TABLE",
    profilePath: process.env.SNOWFLAKE_PROFILE || "profile.json",
  });
}

function sampleRow(eventId) {
  return {
    EVENT_ID: eventId,
    C1: eventId,
    C2: `event-${eventId}`,
  };
}

function track(promise) {
  // Observe rejection immediately so it cannot become an unhandled rejection.
  const pending = { settled: false, result: null };
  pending.outcome = Promise.resolve(promise).then(
    () => ({ ok: true }),
    (error) => ({ ok: false, error }),
  ).then((result) => {
    pending.settled = true;
    pending.result = result;
    return result;
  });
  return pending;
}

async function waitAndRemoveConfirmedPrefix(pending) {
  await pending[0].outcome;
  let confirmed = 0;
  while (pending[0]?.settled) {
    const item = pending.shift();
    if (!item.result.ok) throw item.result.error;
    confirmed++;
  }
  return confirmed;
}

async function main(clientFactory = createClient) {
  const total = Number(process.env.SNOWFLAKE_TEST_ROWS || 10_000);
  const client = await clientFactory();
  const pending = [];
  let confirmed = 0;
  let completed = false;
  try {
    const channel = await client.getElasticChannel();
    for (let eventId = 0; eventId < total; eventId++) {
      // No callback is registered, so the returned Promise identifies the append.
      pending.push(track(channel.appendRowWithWait(sampleRow(eventId), null)));
      if (pending.length >= MAX_PENDING_EVENTS) {
        // Pause intake until the oldest acknowledgement releases one slot.
        confirmed += await waitAndRemoveConfirmedPrefix(pending);
      }
    }

    // End of input: wait until every accepted append is durably acknowledged.
    while (pending.length) {
      confirmed += await waitAndRemoveConfirmedPrefix(pending);
    }
    completed = true;
    console.log(`Durably acknowledged ${confirmed} rows`);
  } finally {
    await client.close({ waitForFlush: completed, timeoutMs: 60_000 });
  }
}

if (require.main === module) {
  const keepAlive = setInterval(() => {}, 1_000);
  main()
    .catch((error) => {
      console.error(error.message);
      process.exitCode = 1;
    })
    .finally(() => clearInterval(keepAlive));
}

module.exports = {
  MAX_PENDING_EVENTS,
  createClient,
  main,
  sampleRow,
  track,
  waitAndRemoveConfirmedPrefix,
};
